### PR TITLE
Improved relay request confirmation mechanism by adding retries

### PR DESCRIPTION
**Pull request by pdyraga**
Before we start relay entry signing process we need to confirm the current relay request start block on the chain. This is to avoid having the client participating in an old relay request signing that has already completed with the rest of the signing group member clients and the result has been already published to the chain.

Such a situation may happen when the current client received multiple blocks at once after a long delay and in those blocks to relay request events to the same signing group were emitted.

The confirmation mechanism has built-in retries. We can retry in case of an error but also when the expected request start block does not match the one currently registered on the chain. Such a situation may happen for Infura-like setup when two or more chain clients are behind a load balancer and they do not have their state in sync yet.

### CODE REVIEW

**Comment by pdyraga** [comment]
The tests nicely illustrate all the possible scenarios:

```
=== RUN   TestConfirmRelayRequestOnFirstAttempt
--- PASS: TestConfirmRelayRequestOnFirstAttempt (0.00s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.044s
```

```
=== RUN   TestConfirmRelayRequestOnLastAttempt
11:36:54.871  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:54.974  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.076  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.177  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.279  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.384  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.485  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.586  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
11:36:55.687  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:255
--- PASS: TestConfirmRelayRequestOnLastAttempt (0.92s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.965s
```

```
=== RUN   TestConfirmRelayRequestOnLastAttemptBecauseOfErrors
11:38:33.315 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.419 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.524 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.625 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.726 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.828 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:33.933 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:34.036 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:38:34.136 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
--- PASS: TestConfirmRelayRequestOnLastAttemptBecauseOfErrors (0.92s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.968s
```

```
=== RUN   TestConfirmRelayRequestWithMoreRecentEntryPending
11:39:21.945  INFO keep-beaco: the currently pending relay request started at block [1889]; skipping the execution of the old relay request from block [1888] beacon.go:240
--- PASS: TestConfirmRelayRequestWithMoreRecentEntryPending (0.00s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.046s
```

```
=== RUN   TestConfirmRelayRequestWithAllAttemptsFailed
11:42:35.998 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.100 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.204 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.306 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.411 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.516 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.621 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.721 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.823 WARNI keep-beaco: could not check current request start block: [VERY BAD ERROR]; will retry after [100ms] beacon.go:220
11:42:36.923 ERROR keep-beaco: could not check current request start block: [VERY BAD ERROR]; giving up after [10] retries beacon.go:211
--- PASS: TestConfirmRelayRequestWithAllAttemptsFailed (0.92s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.971s
```

```
=== RUN   TestConfirmRelayRequestWithAllAttemptsFailed
11:45:18.259  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.362  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.466  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.567  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.669  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.774  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.876  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:18.980  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:19.083  INFO keep-beaco: received unexpected pending relay request start block [0] while the expected was [1888]; will retry after [100ms] beacon.go:258
11:45:19.183 ERROR keep-beaco: could not confirm the expected relay request starting block; the most recent one obtained from chain is [0] and the expected one is [1888]; giving up after [10] retries beacon.go:248
--- PASS: TestConfirmRelayRequestWithAllAttemptsFailed (0.92s)
PASS
ok      github.com/keep-network/keep-core/pkg/beacon    0.970s
```

File: pkg/beacon/beacon.go [hunk 1]
```diff
@@ -76,65 +78,58 @@ func Initialize(
 	node.ResumeSigningIfEligible(relayChain, signing)
 
 	relayChain.OnRelayEntryRequested(func(request *event.Request) {
-		currentEntryStartBlock, err := relayChain.CurrentRequestStartBlock()
-		if err != nil {
-			logger.Warningf("could not check current request start block")
-			return
-		}
+		onConfirmed := func() {
+			if node.IsInGroup(request.GroupPublicKey) {
```
Comment (nkuba) [inline pkg/beacon/beacon.go:82]: Should we check if member is in the group here or before calling `confirmCurrentRelayRequest`?

File: pkg/beacon/beacon.go [hunk 2]
```diff
@@ -199,3 +194,84 @@ func Initialize(
...
+				"could not check current request start block: [%v]; "+
+					"will retry after [%v]",
+				err,
+				delay,
+			)
+			time.Sleep(delay)
+			continue
+		}
+
+		current := currentRequestStartBlock.Uint64()
```
Comment (nkuba) [inline pkg/beacon/beacon.go:243]: Thoughts on adding more context to the `current` variable name?
```suggestion
		currentRequestStartBlock := currentRequestStartBlockBigInt.Uint64()
```

File: pkg/beacon/beacon.go (hunk 1 above)
Comment (pdyraga) [reply pkg/beacon/beacon.go:82]: I don't think so. If the node is not in a signing group it sill enters the signing in a courteous message relay mode. We need it until we update the client to the most recent libp2p with the fix @lukasz-zimnoch implemented for pubsub relays. We do not want the client to enter this mode for old relay entries. Also, no matter if the node is in the group or not, it starts monitoring the chain for relay entry timeouts. Similarly, we do not want the node to do that if the event received is not for the most recent relay request.

**Comment by pdyraga** [comment]
I regenerated contracts bindings, some of them were out of date.

**Pull request merged**